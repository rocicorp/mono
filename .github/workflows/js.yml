name: JS

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

concurrency:
  group: '${{ github.workflow }} @ ${{ github.event.pull_request.head.label || github.head_ref || github.ref }}'
  cancel-in-progress: true

jobs:
  setup:
    name: Setup Dependencies
    runs-on: ubuntu-latest
    env:
      TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
      TURBO_TEAM: rocicorp
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20.x
          cache: 'npm'

      - name: Cache node_modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            node_modules
            node_modules/.vite-temp
          key: npm-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            npm-
      - name: Install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci

  format:
    name: Prettier
    runs-on: ubuntu-latest
    needs: setup
    env:
      TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
      TURBO_TEAM: rocicorp
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20.x
          cache: 'npm'

      - name: Restore cached node_modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            node_modules
            node_modules/.vite-temp
          key: npm-${{ hashFiles('package-lock.json') }}
      - name: Install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
      - run: npm run check-format

  lint:
    name: Lint
    runs-on: ubuntu-latest
    needs: setup
    env:
      TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
      TURBO_TEAM: rocicorp
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20.x
          cache: 'npm'

      - name: Restore cached node_modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            node_modules
            node_modules/.vite-temp
          key: npm-${{ hashFiles('package-lock.json') }}
      - name: Install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
      - run: npm run lint

  check-types:
    name: Check Types
    runs-on: ubuntu-latest
    needs: setup
    env:
      TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
      TURBO_TEAM: rocicorp
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20.x
          cache: 'npm'

      - name: Restore cached node_modules
        uses: actions/cache@v3
        id: cache
        with:
          path: |
            node_modules
            node_modules/.vite-temp
          key: npm-${{ hashFiles('package-lock.json') }}
      - name: Install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
      - run: npm run check-types

  test:
    name: Test
    runs-on: ubuntu-22.04
    needs: setup
    env:
      TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
      TURBO_TEAM: rocicorp
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20.x
          cache: 'npm'

      - name: Restore cached node_modules
        id: cache
        uses: actions/cache@v3
        with:
          path: |
            node_modules
            node_modules/.vite-temp
          key: npm-${{ hashFiles('package-lock.json') }}
      - name: Install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci

      - name: Cache Playwright
        id: cache-playwright
        uses: actions/cache@v3
        with:
          path: ~/.cache/ms-playwright
          key: playwright-${{ runner.os }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            playwright-${{ runner.os }}-
      - name: Install Playwright
        run: npx playwright install --with-deps

      - run: npm run test
