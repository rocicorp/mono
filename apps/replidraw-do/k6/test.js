import ws from "k6/ws";
import { check } from "k6";

const numShapesPerClient = __ENV.SHAPES_PER_CLIENT || 1;
const roomID = __ENV.ROOM_ID;

if (!roomID) {
  throw new Error("Must specify a ROOM_ID env variable");
}

function randomID() {
  return Math.random().toString(36).substring(2);
}

let lastMutationID = 0;

function sendMutation(socket, name, args) {
  const mutation = {
    id: ++lastMutationID,
    name,
    args,
    timestamp: Date.now(),
  };
  const pushBody = {
    mutations: [mutation],
    pushVersion: 1,
    schemaVersion: "",
    timestamp: Date.now(),
  };
  const msg = JSON.stringify(["push", pushBody]);
  console.info("sending", msg);
  socket.send(msg);
}

function createShape(socket, clientID, idx) {
  sendMutation(socket, "createShape", {
    id: `${clientID}-${idx}`,
    shape: randomShape(),
  });
}

function scanShape(socket, clientID, idx) {
  sendMutation(socket, "scanShape", {
    id: `${clientID}-${idx}`,
    dx: 1,
    maxX: 500,
  });
}

function createShapes(socket, clientID, numShapes) {
  for (let i = 0; i < numShapes; i++) {
    createShape(socket, clientID, i);
  }
}

function scanShapes(socket, clientID, numShapes) {
  for (let i = 0; i < numShapes; i++) {
    scanShape(socket, clientID, i);
  }
}

export default function () {
  const clientID = randomID();
  const userID = randomID();

  const socketBaseURL = "ws://127.0.0.1:8787/connect";
  const url = `${socketBaseURL}?clientID=${clientID}&roomID=${roomID}&baseCookie=0&ts=${Date.now()}`;
  const params = {
    headers: {
      "Sec-WebSocket-Protocol": JSON.stringify({
        roomID,
        userID,
      }),
    },
  };

  const response = ws.connect(url, params, function (socket) {
    socket.on("open", function open() {
      console.log("opened");
    });

    socket.on("message", function (message) {
      console.log(`Received message: ${message}`);
      const [type, body] = JSON.parse(message);
      if (type === "connected") {
        createShapes(socket, clientID, numShapesPerClient);
        socket.setInterval(() => {
          scanShapes(socket, clientID, numShapesPerClient);
        }, 16);
      }
    });

    socket.on("close", function () {
      console.log("disconnected");
    });

    socket.on("error", function (e) {
      if (e.error() != "websocket: close sent") {
        console.log("An unexpected error occured: ", e.error());
      }
    });

    socket.setTimeout(function () {
      console.log("2 seconds passed, closing the socket");
      socket.close();
    }, 30000);
  });

  check(response, { "status is 101": (r) => r && r.status === 101 });
}

export function randomShape() {
  const colors = ["red", "blue", "white", "green", "yellow"];
  const s = randInt(100, 400);
  const fill = colors[randInt(0, colors.length - 1)];
  return {
    type: "rect",
    x: randInt(0, 400),
    y: randInt(0, 400),
    width: s,
    height: s,
    rotate: randInt(0, 359),
    fill,
  };
}

export function randInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min); //The maximum is inclusive and the minimum is inclusive
}
