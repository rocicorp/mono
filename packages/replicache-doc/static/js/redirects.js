const redirects = [
  [
    '/byob/remote-database#supabase-setup',
    'https://github.com/rocicorp/todo-nextjs',
  ],
  ['/howto/licensing', '/concepts/licensing'],
  ['/faq#how-does-the-client-know-when-to-sync-does-it-poll', '/byob/poke'],
  [
    '/concepts/faq#what-if-i-dont-have-a-dedicated-backend-i-use-serverless-functions-for-my-backend',
    'https://github.com/rocicorp/todo-nextjs',
  ],
  [
    '/concepts/faq#how-can-i-programmatically-prevent-replicache-from-syncing',
    '/howto/unit-test',
  ],
  [
    '/concepts/faq#unpushed',
    '/api/classes/Replicache#experimentalpendingmutations',
  ],
  ['/concepts/faq#do-you-support-collaborative-text-editing', '/howto/text'],
  [
    '/concepts/faq#what-is-a-monthly-active-profile',
    '/concepts/licensing#monthly-active-profiles',
  ],
  [
    '/concepts/faq#what-do-you-mean-by-commercial-application',
    '/concepts/licensing#pricing-exemption',
  ],
  [
    '/concepts/faq#can-you-give-me-some-billing-examples',
    '/concepts/licensing#pricing-examples',
  ],
  ['/concepts/faq#can-i-get-access-to-the-source-code', '/howto/source-access'],
];

const fullPath = [location.pathname, location.search, location.hash].join('');
for (const [from, to] of redirects) {
  if (fullPath === from) {
    location.href = to;
    break;
  }
}
