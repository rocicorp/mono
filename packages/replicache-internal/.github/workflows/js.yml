name: JS

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  format:
    name: ESLint & Prettier
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 16.x
          cache: 'npm'
      - name: configure npm auth token
        run: npm config set '//registry.npmjs.org/:_authToken' "${NPM_TOKEN}"
        env:
          NPM_TOKEN: ${{secrets.NPM_TOKEN}}
      - run: npm install
      - run: npm run check-format
      - run: npm run lint

  build:
    name: Build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 16.x
          cache: 'npm'
      - name: configure npm auth token
        run: npm config set '//registry.npmjs.org/:_authToken' "${NPM_TOKEN}"
        env:
          NPM_TOKEN: ${{secrets.NPM_TOKEN}}
      - run: npm install
      - run: npm run check-types
      - run: npm run build

  test:
    name: Test
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 16.x
          cache: 'npm'

      - name: configure npm auth token
        run: npm config set '//registry.npmjs.org/:_authToken' "${NPM_TOKEN}"
        env:
          NPM_TOKEN: ${{secrets.NPM_TOKEN}}

      - run: npm ci

      - name: Install Playwright Deps
        run: npx playwright install --with-deps

      - run: npm test
      - run: DD31=true npm test
