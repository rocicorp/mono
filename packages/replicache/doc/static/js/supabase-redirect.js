if (
  location.pathname === '/byob/remote-database' &&
  location.hash === '#supabase-setup'
) {
  location.href = 'https://github.com/rocicorp/todo-nextjs';
}
