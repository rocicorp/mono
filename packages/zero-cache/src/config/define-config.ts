/**
 * Developers can define their configuration via typescript.
 * These types represent the shape that their config must adhere to
 * so we can compile it to a JSON ZeroConfig.
 */
import {normalizeSchema} from '../../../zero-client/src/client/normalized-schema.js';
import type {AST} from '../../../zql/src/zql/ast/ast.js';
import type {Query, SchemaToRow} from '../../../zql/src/zql/query/query.js';
import type {TableSchema} from '../../../zql/src/zql/query/schema.js';
import {ConfigQuery} from './config-query.js';
import {authDataRef, preMutationRowRef} from './refs.js';
import type {
  Action,
  AssetAuthorization as CompiledAssetAuthorization,
  AuthorizationConfig as CompiledAuthorizationConfig,
  ZeroConfig as CompiledZeroConfig,
  EnvRef,
  ZeroConfigSansAuthorization,
} from './zero-config.js';

type Schema = {
  readonly version: number;
  readonly tables: {readonly [table: string]: TableSchema};
};

export type Queries<TSchema extends Schema> = {
  [K in keyof TSchema['tables']]: Query<TSchema['tables'][K]>;
};

type InstanceAuthzRule<TAuthDataShape, TSchema extends TableSchema> = (
  authData: TAuthDataShape,
  row: SchemaToRow<TSchema>,
) => Query<TableSchema>;

type StaticAuthzRule<TAuthDataShape> = (
  authData: TAuthDataShape,
) => Query<TableSchema>;

type StaticAssetAuthorization<TAuthDataShape> = {
  [K in Action]?: StaticAuthzRule<TAuthDataShape>[];
};

type InstanceAssetAuthorization<TAuthDataShape, TSchema extends TableSchema> = {
  [K in Action]?: InstanceAuthzRule<TAuthDataShape, TSchema>[];
};

export type AuthorizationConfig<TAuthDataShape, TSchema extends Schema> = {
  [K in keyof TSchema['tables']]?: {
    table?: StaticAssetAuthorization<TAuthDataShape>;
    column?: {
      [C in keyof TSchema['tables'][K]['columns']]?: StaticAssetAuthorization<TAuthDataShape>;
    };
    row?: InstanceAssetAuthorization<TAuthDataShape, TSchema['tables'][K]>;
    cell?: {
      [C in keyof TSchema['tables'][K]['columns']]?: InstanceAssetAuthorization<
        TAuthDataShape,
        TSchema['tables'][K]
      >;
    };
  };
};

export type ZeroConfig<
  TAuthDataShape,
  TSchema extends Schema,
> = ZeroConfigSansAuthorization & {
  authorization?: AuthorizationConfig<TAuthDataShape, TSchema>;
};

export function runtimeEnv(key: string): EnvRef {
  return {tag: 'env', name: key};
}

export function defineConfig<TAuthDataShape, TSchema extends Schema>(
  schema: TSchema,
  definer: (queries: Queries<TSchema>) => ZeroConfig<TAuthDataShape, TSchema>,
): CompiledZeroConfig {
  const normalizedSchema = normalizeSchema(schema);
  const queries = {} as Record<string, Query<TableSchema>>;
  for (const [name, tableSchema] of Object.entries(normalizedSchema.tables)) {
    queries[name] = new ConfigQuery(tableSchema);
  }

  const config = definer(queries as Queries<TSchema>);
  return compileConfig(config);
}

const DEFAULT_SHARD_ID = '0';
function compileConfig<TAuthDataShape, TSchema extends Schema>(
  config: ZeroConfig<TAuthDataShape, TSchema>,
): CompiledZeroConfig {
  return {
    ...config,
    authorization: compileAuthorization(config.authorization),
    shard: {
      id: config.shard?.id ?? DEFAULT_SHARD_ID,
      publications: config?.shard?.publications ?? [],
    },
  };
}

function compileAuthorization<TAuthDataShape, TSchema extends Schema>(
  authz: AuthorizationConfig<TAuthDataShape, TSchema> | undefined,
): CompiledAuthorizationConfig | undefined {
  if (!authz) {
    return undefined;
  }
  const ret: CompiledAuthorizationConfig = {};
  for (const [tableName, tableConfig] of Object.entries(authz)) {
    ret[tableName] = {
      table: compileTableConfig(tableConfig.table),
      column: compileColumnConfig(tableConfig.column),
      row: compileRowConfig(tableConfig.row),
      cell: compileCellConfig(tableConfig.cell),
    };
  }

  return ret;
}

function compileTableConfig<TAuthDataShape>(
  tableRules: StaticAssetAuthorization<TAuthDataShape> | undefined,
): CompiledAssetAuthorization | undefined {
  if (!tableRules) {
    return undefined;
  }
  return {
    select: compileStaticRules(tableRules.select),
    insert: compileStaticRules(tableRules.insert),
    update: compileStaticRules(tableRules.update),
    delete: compileStaticRules(tableRules.delete),
  };
}

function compileStaticRules<TAuthDataShape>(
  rules: StaticAuthzRule<TAuthDataShape>[] | undefined,
): ['allow', AST][] | undefined {
  if (!rules) {
    return undefined;
  }
  return rules.map(
    rule =>
      [
        'allow',
        (
          rule(authDataRef as TAuthDataShape) as unknown as {
            ast: AST;
          }
        ).ast,
      ] as const,
  );
}

function compileColumnConfig<TAuthDataShape>(
  columnRules:
    | Record<string, StaticAssetAuthorization<TAuthDataShape>>
    | undefined,
): Record<string, CompiledAssetAuthorization> | undefined {
  if (!columnRules) {
    return undefined;
  }
  const ret: Record<string, CompiledAssetAuthorization> = {};
  for (const [columnName, rules] of Object.entries(columnRules)) {
    ret[columnName] = {
      select: compileStaticRules(rules.select),
      insert: compileStaticRules(rules.insert),
      update: compileStaticRules(rules.update),
      delete: compileStaticRules(rules.delete),
    };
  }
  return ret;
}

function compileRowConfig<TAuthDataShape, TSchema extends TableSchema>(
  rowRules: InstanceAssetAuthorization<TAuthDataShape, TSchema> | undefined,
): CompiledAssetAuthorization | undefined {
  if (!rowRules) {
    return undefined;
  }
  return {
    select: compileInstanceRules(rowRules.select),
    insert: compileInstanceRules(rowRules.insert),
    update: compileInstanceRules(rowRules.update),
    delete: compileInstanceRules(rowRules.delete),
  };
}

function compileInstanceRules<TAuthDataShape, TSchema extends TableSchema>(
  rules: InstanceAuthzRule<TAuthDataShape, TSchema>[] | undefined,
): ['allow', AST][] | undefined {
  if (!rules) {
    return undefined;
  }

  return rules.map(
    rule =>
      [
        'allow',
        (
          rule(
            authDataRef as TAuthDataShape,
            preMutationRowRef as SchemaToRow<TSchema>,
          ) as unknown as {
            ast: AST;
          }
        ).ast,
      ] as const,
  );
}

function compileCellConfig<TAuthDataShape, TSchema extends TableSchema>(
  cellRules:
    | Record<string, InstanceAssetAuthorization<TAuthDataShape, TSchema>>
    | undefined,
): Record<string, CompiledAssetAuthorization> | undefined {
  if (!cellRules) {
    return undefined;
  }
  const ret: Record<string, CompiledAssetAuthorization> = {};
  for (const [columnName, rules] of Object.entries(cellRules)) {
    ret[columnName] = {
      select: compileInstanceRules(rules.select),
      insert: compileInstanceRules(rules.insert),
      update: compileInstanceRules(rules.update),
      delete: compileInstanceRules(rules.delete),
    };
  }
  return ret;
}
