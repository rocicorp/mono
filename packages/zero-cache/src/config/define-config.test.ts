import fs from 'fs';
import {afterEach} from 'node:test';
import {beforeEach, expect, type MockInstance, test, vi} from 'vitest';
import {defineConfig, type Queries} from './define-config.js';

type AuthData = {sub: string};

const userSchema = {
  tableName: 'user',
  columns: {
    id: {type: 'string'},
    login: {type: 'string'},
    name: {type: 'string'},
    avatar: {type: 'string'},
    role: {type: 'string'},
  },
  primaryKey: ['id'],
  relationships: {},
} as const;

const issueSchema = {
  tableName: 'issue',
  columns: {
    id: {type: 'string'},
    title: {type: 'string'},
    open: {type: 'boolean'},
    modified: {type: 'number'},
    created: {type: 'number'},
    creatorID: {type: 'string'},
    description: {type: 'string'},
    labelIDs: {type: 'string'},
  },
  primaryKey: ['id'],
  relationships: {
    labels: {
      source: 'id',
      junction: {
        schema: () => issueLabelSchema,
        sourceField: 'issueID',
        destField: 'labelID',
      },
      dest: {
        field: 'id',
        schema: () => labelSchema,
      },
    },
    comments: {
      source: 'id',
      dest: {
        field: 'issueID',
        schema: () => commentSchema,
      },
    },
    creator: {
      source: 'creatorID',
      dest: {
        field: 'id',
        schema: () => userSchema,
      },
    },
  },
} as const;

const commentSchema = {
  tableName: 'comment',
  columns: {
    id: {type: 'string'},
    issueID: {type: 'string'},
    created: {type: 'number'},
    body: {type: 'string'},
    creatorID: {type: 'string'},
  },
  primaryKey: ['id'],
  relationships: {
    creator: {
      source: 'creatorID',
      dest: {
        field: 'id',
        schema: () => userSchema,
      },
    },
  },
} as const;

const labelSchema = {
  tableName: 'label',
  columns: {
    id: {type: 'string'},
    name: {type: 'string'},
  },
  primaryKey: ['id'],
  relationships: {},
} as const;

const issueLabelSchema = {
  tableName: 'issueLabel',
  columns: {
    id: {type: 'string'},
    issueID: {type: 'string'},
    labelID: {type: 'string'},
  },
  primaryKey: ['id'],
  relationships: {},
} as const;

export const schema = {
  version: 1,
  tables: {
    user: userSchema,
    issue: issueSchema,
    comment: commentSchema,
    label: labelSchema,
    issueLabel: issueLabelSchema,
  },
} as const;

const baseConfig = {
  upstreamUri: '',
  cvrDbUri: '',
  changeDbUri: '',
  replicaDbFile: '',
  replicaId: '',
  log: {
    level: 'info',
  },
} as const;

let writeFileSyncMock: MockInstance;
beforeEach(() => {
  writeFileSyncMock = vi
    .spyOn(fs, 'writeFileSync')
    .mockImplementation(() => {});
});

afterEach(() => {
  writeFileSyncMock.mockRestore();
});

test('static authorization rules', () => {
  const config = {
    ...baseConfig,
    authorization: {
      user: {
        table: {
          select: [],
          insert: [],
          update: [],
          delete: [],
        },
        column: {
          id: {
            select: [],
            insert: [],
            update: [],
            delete: [],
          },
        },
      },
    },
  };
  defineConfig(schema, () => config);

  expect(writeFileSyncMock.mock.calls[0][1]).toEqual(
    JSON.stringify(config, null, 2),
  );

  defineConfig(schema, queries => ({
    ...baseConfig,
    authorization: {
      user: {
        table: {
          select: [() => queries.user],
          insert: [() => queries.user],
          update: [() => queries.user],
          delete: [() => queries.user],
        },
        column: {
          id: {
            select: [() => queries.user],
            insert: [() => queries.user],
            update: [() => queries.user],
            delete: [() => queries.user],
          },
        },
      },
    },
  }));

  expect(writeFileSyncMock.mock.calls[1][1]).toMatchSnapshot();

  const policy = (queries: Queries<typeof schema>) => [
    (authData: AuthData) =>
      queries.user.where('id', '=', authData.sub).where('role', '=', 'crew'),
  ];
  defineConfig<AuthData, typeof schema>(schema, queries => ({
    ...baseConfig,
    authorization: {
      user: {
        table: {
          select: policy(queries),
          insert: policy(queries),
          update: policy(queries),
          delete: policy(queries),
        },
        column: {
          login: {
            select: policy(queries),
            insert: policy(queries),
            update: policy(queries),
            delete: policy(queries),
          },
        },
      },
    },
  }));

  expect(writeFileSyncMock.mock.calls[2][1]).toMatchSnapshot();
});

test('instance authorization rules', () => {
  defineConfig<AuthData, typeof schema>(schema, queries => ({
    ...baseConfig,
    authorization: {
      issue: {
        row: {
          update: [
            (authData, row) =>
              queries.issue
                .where('id', '=', row.id)
                .where('creatorID', '=', authData.sub),
            (authData, _row) =>
              queries.user
                .where('id', '=', authData.sub)
                .where('role', '=', 'crew'),
          ],
        },
      },
      comment: {
        row: {
          update: [
            (authData, row) =>
              queries.comment
                .where('id', '=', row.id)
                .where('creatorID', '=', authData.sub),
          ],
        },
        cell: {
          creatorID: {
            update: [
              (authData, _row) =>
                queries.user
                  .where('id', '=', authData.sub)
                  .where('role', '=', 'crew'),
            ],
          },
        },
      },
    },
  }));

  expect(writeFileSyncMock.mock.calls[0][1]).toMatchSnapshot();
});
