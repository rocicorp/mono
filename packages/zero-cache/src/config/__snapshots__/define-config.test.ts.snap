// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`instance authorization rules 1`] = `
"{
  "upstreamUri": "",
  "cvrDbUri": "",
  "changeDbUri": "",
  "replicaDbFile": "",
  "replicaId": "",
  "log": {
    "level": "info"
  },
  "authorization": {
    "issue": {
      "row": {
        "update": [
          [
            "allow",
            {
              "table": "issue",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "preMutationRow",
                    "field": "id"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "creatorID",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                }
              ]
            }
          ],
          [
            "allow",
            {
              "table": "user",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "role",
                  "value": "crew"
                }
              ]
            }
          ]
        ]
      }
    },
    "comment": {
      "row": {
        "update": [
          [
            "allow",
            {
              "table": "comment",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "preMutationRow",
                    "field": "id"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "creatorID",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                }
              ]
            }
          ]
        ]
      },
      "cell": {
        "creatorID": {
          "update": [
            [
              "allow",
              {
                "table": "user",
                "where": [
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "id",
                    "value": {
                      "type": "static",
                      "anchor": "authData",
                      "field": "sub"
                    }
                  },
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "role",
                    "value": "crew"
                  }
                ]
              }
            ]
          ]
        }
      }
    }
  }
}"
`;

exports[`static authorization rules 1`] = `
"{
  "upstreamUri": "",
  "cvrDbUri": "",
  "changeDbUri": "",
  "replicaDbFile": "",
  "replicaId": "",
  "log": {
    "level": "info"
  },
  "authorization": {
    "user": {
      "table": {
        "select": [
          [
            "allow",
            {
              "table": "user"
            }
          ]
        ],
        "insert": [
          [
            "allow",
            {
              "table": "user"
            }
          ]
        ],
        "update": [
          [
            "allow",
            {
              "table": "user"
            }
          ]
        ],
        "delete": [
          [
            "allow",
            {
              "table": "user"
            }
          ]
        ]
      },
      "column": {
        "id": {
          "select": [
            [
              "allow",
              {
                "table": "user"
              }
            ]
          ],
          "insert": [
            [
              "allow",
              {
                "table": "user"
              }
            ]
          ],
          "update": [
            [
              "allow",
              {
                "table": "user"
              }
            ]
          ],
          "delete": [
            [
              "allow",
              {
                "table": "user"
              }
            ]
          ]
        }
      }
    }
  }
}"
`;

exports[`static authorization rules 2`] = `
"{
  "upstreamUri": "",
  "cvrDbUri": "",
  "changeDbUri": "",
  "replicaDbFile": "",
  "replicaId": "",
  "log": {
    "level": "info"
  },
  "authorization": {
    "user": {
      "table": {
        "select": [
          [
            "allow",
            {
              "table": "user",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "role",
                  "value": "crew"
                }
              ]
            }
          ]
        ],
        "insert": [
          [
            "allow",
            {
              "table": "user",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "role",
                  "value": "crew"
                }
              ]
            }
          ]
        ],
        "update": [
          [
            "allow",
            {
              "table": "user",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "role",
                  "value": "crew"
                }
              ]
            }
          ]
        ],
        "delete": [
          [
            "allow",
            {
              "table": "user",
              "where": [
                {
                  "type": "simple",
                  "op": "=",
                  "field": "id",
                  "value": {
                    "type": "static",
                    "anchor": "authData",
                    "field": "sub"
                  }
                },
                {
                  "type": "simple",
                  "op": "=",
                  "field": "role",
                  "value": "crew"
                }
              ]
            }
          ]
        ]
      },
      "column": {
        "login": {
          "select": [
            [
              "allow",
              {
                "table": "user",
                "where": [
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "id",
                    "value": {
                      "type": "static",
                      "anchor": "authData",
                      "field": "sub"
                    }
                  },
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "role",
                    "value": "crew"
                  }
                ]
              }
            ]
          ],
          "insert": [
            [
              "allow",
              {
                "table": "user",
                "where": [
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "id",
                    "value": {
                      "type": "static",
                      "anchor": "authData",
                      "field": "sub"
                    }
                  },
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "role",
                    "value": "crew"
                  }
                ]
              }
            ]
          ],
          "update": [
            [
              "allow",
              {
                "table": "user",
                "where": [
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "id",
                    "value": {
                      "type": "static",
                      "anchor": "authData",
                      "field": "sub"
                    }
                  },
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "role",
                    "value": "crew"
                  }
                ]
              }
            ]
          ],
          "delete": [
            [
              "allow",
              {
                "table": "user",
                "where": [
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "id",
                    "value": {
                      "type": "static",
                      "anchor": "authData",
                      "field": "sub"
                    }
                  },
                  {
                    "type": "simple",
                    "op": "=",
                    "field": "role",
                    "value": "crew"
                  }
                ]
              }
            ]
          ]
        }
      }
    }
  }
}"
`;
